Use Lettuce for Orbit Messaging

tmccnnll opened the pull request:
Change Orbit messaging to use [Lettuce](https://github.com/lettuce-io/lettuce-core) instead of Redisson.

Commit:
Rename RedisOrbitClient to RedissonOrbitClient

* Rename class to prepare for addition of LettuceOrbitClient class

Commit:
Update Redisson project dependency from 3.4.4 to 3.5.7

* Update RedisPipelineCodec due to changes in Redisson Encoder/Decoder
* Update RedisPipelineStep and its implementations to align with changes to RedisPipelineCodec (use of ByteBuf instead of byte array)
* Add unit tests for RedisPipelineCodec and RedisCompressionPipelineStep

Commit:
Add  equals, hashCode methods and new ctor to RedisMsg

* Add unit tests for equals and hashCode methods in RedisMsg

Commit:
Add LettuceOrbitClient and use it for Orbit messaging

* Add Lettuce and FST as a project dependencies
* Create LettuceOrbitClass to handle pub/sub Redis work
* Create FstSerializedObjectCodec to be used with LettuceOrbitClient
* Create unit test for FstSerializedObjectCodec
* Change RedisConnectionManager#messagingClients to be list of LettuceOrbitClient
* Change RedisClusterPeer#join(...) to use Lettuce compatible listener

Commit:
Refactor Redisson and Lettuce specific classes into their own packages

Commit:
Increase version of orbit-core dependency from 1.3.0 to 1.7.0

Commit:
Bump version number from 1.2.8-SNAPSHOT to 1.2.9-SNAPSHOT

ceckhardt commented on pom.xml:
```diff
@@ -39,11 +39,11 @@ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
     <groupId>cloud.orbit</groupId>
     <artifactId>orbit-redis-cluster</artifactId>
-    <version>1.2.8-SNAPSHOT</version>
+    <version>1.2.9-SNAPSHOT</version>
```
So this should likely be 1.2.8.4-SNAPSHOT since this was forked from 1.2.8 and we already have a 1.2.8.3

tmccnnll replied:
Makes sense. I'm going to see what Joe says about the preferred version number in the PR I have open in the base project before making any changes.

Commit:
Update LettuceOrbitClient to separate subscribing and publishing

Accidentally missed this change when putting together branch for PR.

jpennell commented on src/main/java/cloud/orbit/actors/cluster/impl/lettuce/LettuceOrbitClient.java:
```diff
@@ -0,0 +1,101 @@
+        else
+        {
+            logger.error("Error subscribing to channel [{}]", channelId);
+            return new CompletableFuture<>();
```
I believe this `CompletableFuture` will never complete. 
Might be better to return:
```
CompletableFuture.completedFuture(null)
```
or 
```
final CompletableFuture<Void> result = new CompletableFuture<>();
result.completeExceptionally(new IllegalStateException("Error subscribing to channel...."));
return result;
```
?

tmccnnll replied:
I like the latter approach.

JordanAngold commented on src/main/java/cloud/orbit/actors/cluster/impl/lettuce/FstSerializedObjectCodec.java:
```diff
@@ -0,0 +1,92 @@
+        }
+        catch (Exception e)
+        {
+            return null;
```
Is this exception-discarding behaviour analogous to the existing Redisson implementation?

tmccnnll replied:
Throwing an `IllegalStateException` seems to be the preferred approach. I'm basing this on what I see  in Lettuce's [`CompressionCodec`](https://github.com/lettuce-io/lettuce-core/blob/master/src/main/java/com/lambdaworks/redis/codec/CompressionCodec.java) and [`StringCodec`](https://github.com/lettuce-io/lettuce-core/blob/master/src/main/java/com/lambdaworks/redis/codec/StringCodec.java)

Commit:
Return empty list from RedisBasicPipeline.defaultPipeline()

* Default shouldn't be RedisCompressionPipelineStep
* Custom pipeline steps can still be defined in RedisClusterConfig

Commit:
Remove null return values from FstSerializedObjectCodec

* Throw IllegalStateException instead
* Same pattern as used by Lettuce CompressionCodec, StringCodec

Commit:
Fix return values when pub/sub connections unavailable

Commit:
Improve ByteBuf usage in RedisPipelineCodec#decoder

* Release `conversionBytes` ByteBuf variable to prevent memory leak
* Use ByteBufAllocator instead of Unpooled for creating ByteBuf

Commit:
Rewrite RedisCompressionPipelineStep based off LZ4Codec from Redisson

This pipeline step is essentially doing the same thing as the Redisson provided LZ4Codec so I rewrote it to copy the majority of the LZ4Codec implementation.

The LZ4Codec implementation can be found here: https://github.com/redisson/redisson/blob/master/redisson/src/main/java/org/redisson/codec/LZ4Codec.java

Commit:
Change SerializationCodec to FstCodec in RedisConnectionManager

Redisson clients were being configured to use the Redisson provided SerializationCodec when Java serialization was desired. Now that FST is a dependency of this project, it makes sense to change SerializationCodec to FstCodec since the latter is more performant and is a drop-in replacement for JDK serialization.

Commit:
Reverting change to version number

Temporarily reverting change to version number to allow PR in fork of project to be merged: https://github.com/ceckhardt/orbit-redis-cluster/pull/1
